src/ray/gcs/redis_context.cc

```diff
@@ -357,7 +370,7 @@ Status RedisContext::Connect(const std::string &address, int port, bool sharding
   RAY_CHECK(!redis_async_context_);
   RAY_CHECK(!async_redis_subscribe_context_);
 
-  RAY_CHECK_OK(ConnectWithRetries(address, port, redisConnect, &context_));
+  RAY_CHECK_OK(ConnectOrFail(address, port, redisConnect, &context_));
```

Comment:
`ConnectOrFail ` always returns OK, right?
can it return other Status while connect failed?

Reply:
`ConnectOrFail` is intended to be a drop-in replacement for the old `ConnectWithRetries`, which indeed never actually returned anything other than OK.

Reply:
`ConnectOrFail` always return OK but `ConnectWithRetries` returns OK and error Status. The naming is kind of contrarily, i think.

Reply:
> `ConnectOrFail` always return OK but `ConnectWithRetries` returns OK and error Status. The naming is kind of contrarily, i think.

Uh...you think so? The point of renaming `ConnectWithRetries` to `ConnectOrFail` and giving the name `ConnectWithRetries` to a sub-function was that it seemed kind of unexpected that a function called "ConnectWithRetries" might crash the entire program. The point of the name "ConnectOrFail" is to get across that when you call this function, it might exit.

Is it the word "Fail" that isn't working for you? Maybe "ConnectOrAbort" or "ConnectOrFatal" or something?

> `ConnectOrFail` always return OK but `ConnectWithRetries` returns OK and error Status.

The set of situations where `ConnectOrFail` will return OK is literally exactly the same set of situations where `ConnectWithRetries` returns OK. The only difference in their behavior is what happens if they *don't* return OK.

Reply:
I think from point of naming, methods suffixed with `Fail` should fail(returns status that is not ok) if the condition was not satisfied, while that suffixed with `Retries` should retry for couple of times if the condition was not satisfied for temporarily.

Reply:
Okay, but what we have is a function that exits the program with `RAY_LOG(FATAL)`. I agree that we need clarity between the version that retries and the version that doesn't. We do indeed have ConnectWithRetries versus ConnectWithoutRetries to clarify whether a function retries or not, and it sounds like that is something you want, right? So the question remaining is what to call the function that might exit the program. You're talking about *what* a function returns, but what's special about this function is that it might never return at all. I'm perfectly happy to call it anything, just as long as there is some indication somewhere in there that it might not return. ConnectOrAbort, ConnectOrFatal, ConnectOrExit, anything.

Reply:
Thanks for your try in the tiny naming thing :)
for b658a7f i think we just keep the if branch, like 
`if (status.IsRedisError()) {
....
}`, while the `else` branch is not necessary.

Reply:
> i think we just keep the if branch, like
> `if (status.IsRedisError()) { .... }`, while the `else` branch is not necessary.

Uh, I mean, we can, and that'll be harmless at the moment, but I thought the point was to future-proof against adding other possible errors? The function currently called `ConnectOrFail`  is the heir to the original `ConnectWithRetries` which could never return without connecting successfully. If we change that, then it will be possible for the function to return *without* connecting. I assume that the reason the function was written that way in the first place was because all the code after it relies on the fact that the connection was definitely successful.

It seems like the *safest* thing to do when the function encounters an unanticipated type of error is to halt and catch fire. Allowing the code to just keep going and ignore the failed connection would mean we could get unexpected behavior in the code following the connection call.

Basically, what we started with is:
```
Connect()
assert IsConnected()
(lots more code)
```
It sounds like you're talking about
```
Connect()
assert not IsOneParticularKindOfError()
(lots more code)
```

Reply:
For now `ConnectWithRetries` only returns OK or RedisError, i think in the codes where it's invoked it's ok we only need to check these two status. In the future if more status returned, we can add their handle logic ( which is perhaps a different logic).

Reply:
Like in https://github.com/ray-project/ray/pull/12976?

Reply:
https://github.com/ray-project/ray/pull/12976/files#diff-478a807e3003772df7d270ff7ba820665462d5c82d008dff07dc0f3d4747af92R355
Why just write as like:
```
  if (status.IsRedisError()) {
      RAY_LOG(FATAL) << status.message();
  }
```

Reply:
I think that silently swallowing all other possible errors and returning OK without even a hint that something is happening there is asking for trouble, but I don't feel very strongly about it.
https://github.com/ray-project/ray/pull/12976/commits/66b2bf85a93e6050e5134cfcae708382fd73a3fc

Reply:
Ok. How about this?
```
  if (!status.ok()) {
      RAY_LOG(FATAL) << "Error when connect to redis with status " << status <<  " and error message "   status.message();
  }
```
You can modify the sentence as you want, with printing error code and message :)

Reply:
Oh! I think I completely misunderstood what you were driving at. Are you just concerned that some future `.message()` might not contain enough information? I can use `ToString()`, that should include all the relevant information from the status, right? Does https://github.com/ray-project/ray/pull/12976 look good?

Reply:
LGTM